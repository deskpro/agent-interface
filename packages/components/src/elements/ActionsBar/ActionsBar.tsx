import * as React from "react";
import classNames from "classnames";

import Dropdown from "../../Button/Dropdown";
import Menu from "../Menu/Menu";
import { MenuItemProps } from "../Menu/MenuItem";
import Icon from "../Icon/Icon";

import "@deskpro/agent-interface-style/dist/components/dp-actions.css";

export interface ActionBarItemType {
  name: React.Key;
  title: string;
  icon: string;
  items: MenuItemProps[];
}

export type ActionsBarProps = {
  className?: string;
  type?: "default" | "outlined";
  items: ActionBarItemType[];
  onItemClick: (itemName: React.Key, e: React.SyntheticEvent) => void;
};

const ActionsBar: React.FC<ActionsBarProps> = ({
  className,
  type = "default",
  items,
  onItemClick
}) => (
  <div
    className={classNames("dp-ActionsBar", className, {
      outlined: type === "outlined"
    })}
  >
    {items.map(({ items: subItems, name, icon, title }) => (
      <Dropdown
        key={name}
        renderMenu={menuProps => (
          <Menu {...menuProps}>
            {subItems.map(({ name: menuItemName, ...menuItemProps }) => (
              <Menu.MenuItem
                key={menuItemName}
                name={menuItemName}
                {...menuItemProps}
                onClick={onItemClick}
              />
            ))}
          </Menu>
        )}
      >
        <Icon name={icon} size={type === "outlined" ? 15 : 23} />
        {title}
      </Dropdown>
    ))}
  </div>
);

export default ActionsBar;
