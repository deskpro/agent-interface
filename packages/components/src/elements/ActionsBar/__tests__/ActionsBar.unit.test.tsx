import * as React from "react";
import { render, fireEvent } from "react-testing-library";

import ActionsBar from "../ActionsBar";

const actionsBar = [
  {
    name: "tasks",
    title: "Tasks",
    icon: "tasks",
    items: [
      { name: "lock", text: "Lock", icon: "lock" },
      { name: "hold", text: "Put the ticket on hold", icon: "on-hold" },
      { name: "delete", text: "Delete", icon: "bin" }
    ]
  },
  {
    name: "billing",
    title: "Billing",
    icon: "billing",
    items: [
      { name: "create-linked", text: "Create linked ticket", icon: "lock" },
      { name: "link-existing", text: "Link existing tiket", icon: "link" },
      { name: "split", text: "Split message to new ticket", icon: "transfer" },
      { name: "merge", text: "Merge", icon: "merge" }
    ]
  },
  {
    name: "license",
    title: "License",
    icon: "liscense",
    items: [
      { name: "create-linked", text: "Create linked ticket", icon: "lock" },
      { name: "link-existing", text: "Link existing tiket", icon: "link" },
      { name: "split", text: "Split message to new ticket", icon: "transfer" },
      { name: "merge", text: "Merge", icon: "merge" }
    ]
  },
  {
    name: "ticket-properties",
    title: "Ticket Properties",
    icon: "ticket-properties",
    items: [
      { name: "print", text: "Print", icon: "print" },
      { name: "download-pdf", text: "Download PDF", icon: "download" },
      { name: "download-debug", text: "Download debug file", icon: "beetle" }
    ]
  },
  {
    name: "follow-up",
    title: "Follow-ups",
    icon: "follow-up",
    items: [
      { name: "ban", text: "Ban", icon: "ban" },
      { name: "spam", text: "Spam", icon: "spam" }
    ]
  }
];

describe("<ActionsBar />", () => {
  it("should match snapshot for the general bar", () => {
    const { container } = render(
      <ActionsBar items={actionsBar} onItemClick={jest.fn()} />
    );
    expect(container.firstChild).toMatchSnapshot("General ActionsBar");
  });

  it("should match snapshot for the outlined bar", () => {
    const { container } = render(
      <ActionsBar items={actionsBar} type="outlined" onItemClick={jest.fn()} />
    );
    expect(container.firstChild).toMatchSnapshot("Outlined ActionsBar");
  });

  it("should call item click handler with correct arguments", () => {
    const handleItemClick = jest.fn();
    const { getByText } = render(
      <ActionsBar
        items={actionsBar}
        type="outlined"
        onItemClick={handleItemClick}
      />
    );

    expect(getByText("Tasks")).toBeInTheDocument();
    fireEvent.click(getByText("Tasks"));
    expect(getByText("Lock")).toBeInTheDocument();
    fireEvent.click(getByText("Lock"));
    expect(handleItemClick).toBeCalledTimes(1);
    expect(handleItemClick.mock.calls[0][0]).toBe("lock");
  });
});
