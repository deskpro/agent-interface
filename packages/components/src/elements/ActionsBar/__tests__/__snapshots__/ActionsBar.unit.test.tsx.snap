// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionsBar /> should match snapshot for the general bar: General ActionsBar 1`] = `
<div
  class="dp-ActionsBar"
>
  <span
    class="dp-Dropdown-container"
  >
    <span
      class="dp-Dropdown"
      role="button"
      tabindex="-1"
    >
      <span
        class="dp-Icon Icon--tasks"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="23"
          width="23"
        />
      </span>
      Tasks
      <span
        class="dp-Icon dp-Arrow"
        role="button"
        tabindex="-1"
      />
    </span>
  </span>
  <span
    class="dp-Dropdown-container"
  >
    <span
      class="dp-Dropdown"
      role="button"
      tabindex="-1"
    >
      <span
        class="dp-Icon Icon--billing"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="23"
          width="23"
        />
      </span>
      Billing
      <span
        class="dp-Icon dp-Arrow"
        role="button"
        tabindex="-1"
      />
    </span>
  </span>
  <span
    class="dp-Dropdown-container"
  >
    <span
      class="dp-Dropdown"
      role="button"
      tabindex="-1"
    >
      <span
        class="dp-Icon Icon--liscense"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="23"
          width="23"
        />
      </span>
      License
      <span
        class="dp-Icon dp-Arrow"
        role="button"
        tabindex="-1"
      />
    </span>
  </span>
  <span
    class="dp-Dropdown-container"
  >
    <span
      class="dp-Dropdown"
      role="button"
      tabindex="-1"
    >
      <span
        class="dp-Icon Icon--ticket-properties"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="23"
          width="23"
        />
      </span>
      Ticket Properties
      <span
        class="dp-Icon dp-Arrow"
        role="button"
        tabindex="-1"
      />
    </span>
  </span>
  <span
    class="dp-Dropdown-container"
  >
    <span
      class="dp-Dropdown"
      role="button"
      tabindex="-1"
    >
      <span
        class="dp-Icon Icon--follow-up"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="23"
          width="23"
        />
      </span>
      Follow-ups
      <span
        class="dp-Icon dp-Arrow"
        role="button"
        tabindex="-1"
      />
    </span>
  </span>
</div>
`;

exports[`<ActionsBar /> should match snapshot for the outlined bar: Outlined ActionsBar 1`] = `
<div
  class="dp-ActionsBar outlined"
>
  <span
    class="dp-Dropdown-container"
  >
    <span
      class="dp-Dropdown"
      role="button"
      tabindex="-1"
    >
      <span
        class="dp-Icon Icon--tasks"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="15"
          width="15"
        />
      </span>
      Tasks
      <span
        class="dp-Icon dp-Arrow"
        role="button"
        tabindex="-1"
      />
    </span>
  </span>
  <span
    class="dp-Dropdown-container"
  >
    <span
      class="dp-Dropdown"
      role="button"
      tabindex="-1"
    >
      <span
        class="dp-Icon Icon--billing"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="15"
          width="15"
        />
      </span>
      Billing
      <span
        class="dp-Icon dp-Arrow"
        role="button"
        tabindex="-1"
      />
    </span>
  </span>
  <span
    class="dp-Dropdown-container"
  >
    <span
      class="dp-Dropdown"
      role="button"
      tabindex="-1"
    >
      <span
        class="dp-Icon Icon--liscense"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="15"
          width="15"
        />
      </span>
      License
      <span
        class="dp-Icon dp-Arrow"
        role="button"
        tabindex="-1"
      />
    </span>
  </span>
  <span
    class="dp-Dropdown-container"
  >
    <span
      class="dp-Dropdown"
      role="button"
      tabindex="-1"
    >
      <span
        class="dp-Icon Icon--ticket-properties"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="15"
          width="15"
        />
      </span>
      Ticket Properties
      <span
        class="dp-Icon dp-Arrow"
        role="button"
        tabindex="-1"
      />
    </span>
  </span>
  <span
    class="dp-Dropdown-container"
  >
    <span
      class="dp-Dropdown"
      role="button"
      tabindex="-1"
    >
      <span
        class="dp-Icon Icon--follow-up"
        role="button"
        tabindex="-1"
      >
        <icon-mock
          height="15"
          width="15"
        />
      </span>
      Follow-ups
      <span
        class="dp-Icon dp-Arrow"
        role="button"
        tabindex="-1"
      />
    </span>
  </span>
</div>
`;
